File: sdks/js_sdk/src/runtime/gateway.ts
import { ErrorCode } from '../internal/errorMapping.js';
import type { HandoffRequest, HandoffResponse } from '../clients/handoff.js';

export const REGISTRATION_TYPE_STANDARD_AGENT = 1;
export const REGISTRATION_TYPE_SWARM_GATEWAY = 2;

export const AGENT_STATE_INITIALIZING = 1;
export const AGENT_STATE_RUNNING = 4;
export const AGENT_STATE_FAILED = 10;
export const AGENT_STATE_SHUTTING_DOWN = 11;

export const DEFAULT_PEER_LIVENESS_THRESHOLD_MS = 30_000;

export const NON_SERVING_AGENT_STATES = new Set<number>([
  AGENT_STATE_INITIALIZING,
  AGENT_STATE_FAILED,
  AGENT_STATE_SHUTTING_DOWN,
]);

export interface GatewayPeerDescriptor {
  agentId: string;
  registrationType: number;
  capabilities: string[];
}

export interface PeerRuntimeState {
  state: number;
  lastHeartbeatMs: number;
  cooldownUntilMs: number;
}

export interface PeerSelectorOptions {
  localAgentId: string;
  localCapabilities: string[];
  nowMsFn?: () => number;
  peerHealthFn?: (peer: GatewayPeerDescriptor) => boolean;
  livenessThresholdMs?: number;
}

export interface UpdatePeerRuntimeStateOptions {
  state?: number;
  lastHeartbeatMs?: number;
  cooldownUntilMs?: number;
}

export interface RecordPeerOverloadedOptions {
  retryAfterMs?: number;
  localCooldownMs?: number;
}

export interface GatewayRedirectEmitterOptions {
  agentId: string;
  capabilities?: string[];
  retryAfterMs?: number;
  peerDescriptors?: GatewayPeerDescriptor[];
  peerHealthFn?: (peer: GatewayPeerDescriptor) => boolean;
  nowMsFn?: () => number;
  peerLivenessThresholdMs?: number;
}

export class GatewayValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'GatewayValidationError';
  }
}

function defaultNowMs(): number {
  return Date.now();
}

function sameCapabilities(
  peerCapabilities: string[],
  localCapabilities: Set<string>
): boolean {
  const peerSet = new Set(peerCapabilities);
  if (peerSet.size !== localCapabilities.size) {
    return false;
  }
  for (const localCapability of localCapabilities) {
    if (!peerSet.has(localCapability)) {
      return false;
    }
  }
  return true;
}

export class PeerSelector {
  private readonly localAgentId: string;
  private readonly localCapabilities: Set<string>;
  private readonly nowMs: () => number;
  private readonly peerHealthFn: (peer: GatewayPeerDescriptor) => boolean;
  private readonly livenessThresholdMs: number;

  private peers: GatewayPeerDescriptor[] = [];
  private runtime: Map<string, PeerRuntimeState> = new Map();
  private rrCursor = 0;

  constructor(options: PeerSelectorOptions) {
    const threshold = options.livenessThresholdMs ?? DEFAULT_PEER_LIVENESS_THRESHOLD_MS;
    if (threshold < 0) {
      throw new GatewayValidationError('livenessThresholdMs must be >= 0');
    }

    this.localAgentId = options.localAgentId;
    this.localCapabilities = new Set(options.localCapabilities ?? []);
    this.nowMs = options.nowMsFn ?? defaultNowMs;
    this.peerHealthFn = options.peerHealthFn ?? (() => true);
    this.livenessThresholdMs = threshold;
  }

  setPeers(peers: GatewayPeerDescriptor[]): void {
    this.peers = [...peers];

    const activeIds = new Set(this.peers.map((peer) => peer.agentId));
    for (const agentId of [...this.runtime.keys()]) {
      if (!activeIds.has(agentId)) {
        this.runtime.delete(agentId);
      }
    }

    const nowMs = this.nowMs();
    for (const peer of this.peers) {
      if (!this.runtime.has(peer.agentId)) {
        this.runtime.set(peer.agentId, {
          state: AGENT_STATE_RUNNING,
          lastHeartbeatMs: nowMs,
          cooldownUntilMs: 0,
        });
      }
    }
  }

  updatePeerRuntimeState(
    agentId: string,
    options: UpdatePeerRuntimeStateOptions = {}
  ): void {
    const runtime = this.ensureRuntime(agentId);
    if (options.state !== undefined) {
      runtime.state = options.state;
    }
    if (options.lastHeartbeatMs !== undefined) {
      runtime.lastHeartbeatMs = Math.max(Math.floor(options.lastHeartbeatMs), 0);
    }
    if (options.cooldownUntilMs !== undefined) {
      runtime.cooldownUntilMs = Math.max(Math.floor(options.cooldownUntilMs), 0);
    }
    this.runtime.set(agentId, runtime);
  }

  touchPeerHeartbeat(
    agentId: string,
    options: Pick<UpdatePeerRuntimeStateOptions, 'state'> & { nowMs?: number } = {}
  ): void {
    this.updatePeerRuntimeState(agentId, {
      state: options.state,
      lastHeartbeatMs: options.nowMs ?? this.nowMs(),
    });
  }

  recordPeerOverloaded(
    agentId: string,
    options: RecordPeerOverloadedOptions = {}
  ): void {
    const runtime = this.ensureRuntime(agentId);
    const nowMs = this.nowMs();
    const retryAfterMs = Math.max(Math.floor(options.retryAfterMs ?? 0), 0);
    const localCooldownMs = Math.max(Math.floor(options.localCooldownMs ?? 0), 0);
    const cooldownMs = Math.max(retryAfterMs, localCooldownMs);
    runtime.cooldownUntilMs = Math.max(runtime.cooldownUntilMs, nowMs + cooldownMs);
    this.runtime.set(agentId, runtime);
  }

  isEligible(peer: GatewayPeerDescriptor): boolean {
    if (peer.registrationType !== REGISTRATION_TYPE_SWARM_GATEWAY) {
      return false;
    }
    if (peer.agentId === this.localAgentId) {
      return false;
    }
    return sameCapabilities(peer.capabilities, this.localCapabilities);
  }

  selectPeer(): string {
    const healthy = this.peers.filter((peer) => this.isHealthy(peer));
    if (healthy.length === 0) {
      return '';
    }

    const idx = this.rrCursor % healthy.length;
    this.rrCursor += 1;
    return healthy[idx].agentId;
  }

  private ensureRuntime(agentId: string): PeerRuntimeState {
    const existing = this.runtime.get(agentId);
    if (existing !== undefined) {
      return existing;
    }
    return {
      state: AGENT_STATE_RUNNING,
      lastHeartbeatMs: this.nowMs(),
      cooldownUntilMs: 0,
    };
  }

  private isHealthy(peer: GatewayPeerDescriptor): boolean {
    if (!this.isEligible(peer)) {
      return false;
    }
    if (!this.peerHealthFn(peer)) {
      return false;
    }

    const runtime = this.runtime.get(peer.agentId);
    if (runtime === undefined) {
      return false;
    }

    const nowMs = this.nowMs();
    if (runtime.cooldownUntilMs > nowMs) {
      return false;
    }
    if (nowMs - runtime.lastHeartbeatMs > this.livenessThresholdMs) {
      return false;
    }
    if (NON_SERVING_AGENT_STATES.has(runtime.state)) {
      return false;
    }
    return true;
  }
}

export class GatewayRedirectEmitter {
  private readonly retryAfterMs: number;
  private readonly peerSelector: PeerSelector;

  constructor(options: GatewayRedirectEmitterOptions) {
    if ((options.retryAfterMs ?? 0) < 0) {
      throw new GatewayValidationError('retryAfterMs must be >= 0');
    }

    this.retryAfterMs = options.retryAfterMs ?? 1000;
    this.peerSelector = new PeerSelector({
      localAgentId: options.agentId,
      localCapabilities: options.capabilities ?? [],
      nowMsFn: options.nowMsFn,
      peerHealthFn: options.peerHealthFn,
      livenessThresholdMs: options.peerLivenessThresholdMs,
    });
    this.peerSelector.setPeers(options.peerDescriptors ?? []);
  }

  setPeerDescriptors(peers: GatewayPeerDescriptor[]): void {
    this.peerSelector.setPeers(peers);
  }

  updatePeerRuntimeState(
    agentId: string,
    options: UpdatePeerRuntimeStateOptions = {}
  ): void {
    this.peerSelector.updatePeerRuntimeState(agentId, options);
  }

  touchPeerHeartbeat(
    agentId: string,
    options: Pick<UpdatePeerRuntimeStateOptions, 'state'> & { nowMs?: number } = {}
  ): void {
    this.peerSelector.touchPeerHeartbeat(agentId, options);
  }

  recordPeerOverloaded(
    agentId: string,
    options: RecordPeerOverloadedOptions = {}
  ): void {
    this.peerSelector.recordPeerOverloaded(agentId, options);
  }

  emitOverloadedResponse(request: HandoffRequest): HandoffResponse {
    if (request.delegationPolicy?.allowSpilloverRouting) {
      const redirectToAgentId = this.peerSelector.selectPeer();
      if (redirectToAgentId) {
        return {
          requestId: request.requestId,
          accepted: false,
          rejectionReason: 'Gateway at capacity; redirect to peer gateway',
          rejectionCode: ErrorCode.REDIRECT,
          retryAfterMs: 0,
          redirectToAgentId,
        };
      }
    }

    return {
      requestId: request.requestId,
      accepted: false,
      rejectionReason: 'Gateway at capacity',
      rejectionCode: ErrorCode.OVERLOADED,
      retryAfterMs: this.retryAfterMs,
    };
  }
}

File: sdks/rust_sdk/src/clients/gateway.rs
use crate::constants::agent_state;
use std::collections::{HashMap, HashSet};
use std::sync::Arc;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

use super::handoff::{
    HandoffRequest, HandoffResponse, HandoffStatus, REJECTION_CODE_OVERLOADED,
    REJECTION_CODE_REDIRECT,
};

pub const REGISTRATION_TYPE_STANDARD_AGENT: i32 = 1;
pub const REGISTRATION_TYPE_SWARM_GATEWAY: i32 = 2;
pub const DEFAULT_PEER_LIVENESS_THRESHOLD_MS: u64 = 30_000;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct GatewayPeerDescriptor {
    pub agent_id: String,
    pub registration_type: i32,
    pub capabilities: Vec<String>,
}

impl GatewayPeerDescriptor {
    pub fn new(agent_id: String) -> Self {
        Self {
            agent_id,
            registration_type: REGISTRATION_TYPE_SWARM_GATEWAY,
            capabilities: Vec::new(),
        }
    }

    pub fn with_registration_type(mut self, registration_type: i32) -> Self {
        self.registration_type = registration_type;
        self
    }

    pub fn with_capabilities(mut self, capabilities: Vec<String>) -> Self {
        self.capabilities = capabilities;
        self
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PeerRuntimeState {
    pub state: i32,
    pub last_heartbeat_ms: u64,
    pub cooldown_until_ms: u64,
}

impl PeerRuntimeState {
    fn new(last_heartbeat_ms: u64) -> Self {
        Self {
            state: agent_state::RUNNING,
            last_heartbeat_ms,
            cooldown_until_ms: 0,
        }
    }
}

pub type PeerHealthFn = Arc<dyn Fn(&GatewayPeerDescriptor) -> bool + Send + Sync>;
pub type NowMsFn = Arc<dyn Fn() -> u64 + Send + Sync>;

#[derive(Clone)]
pub struct PeerSelectorOptions {
    pub local_agent_id: String,
    pub local_capabilities: Vec<String>,
    pub now_ms_fn: Option<NowMsFn>,
    pub peer_health_fn: Option<PeerHealthFn>,
    pub liveness_threshold_ms: u64,
}

impl PeerSelectorOptions {
    pub fn new(local_agent_id: String, local_capabilities: Vec<String>) -> Self {
        Self {
            local_agent_id,
            local_capabilities,
            now_ms_fn: None,
            peer_health_fn: None,
            liveness_threshold_ms: DEFAULT_PEER_LIVENESS_THRESHOLD_MS,
        }
    }
}

pub struct PeerSelector {
    local_agent_id: String,
    local_capabilities: HashSet<String>,
    now_ms_fn: NowMsFn,
    peer_health_fn: PeerHealthFn,
    liveness_threshold_ms: u64,
    peers: Vec<GatewayPeerDescriptor>,
    runtime: HashMap<String, PeerRuntimeState>,
    rr_cursor: usize,
}

impl PeerSelector {
    pub fn new(options: PeerSelectorOptions) -> Self {
        Self {
            local_agent_id: options.local_agent_id,
            local_capabilities: options.local_capabilities.into_iter().collect(),
            now_ms_fn: options
                .now_ms_fn
                .unwrap_or_else(|| Arc::new(default_now_ms as fn() -> u64)),
            peer_health_fn: options.peer_health_fn.unwrap_or_else(|| Arc::new(|_| true)),
            liveness_threshold_ms: options.liveness_threshold_ms,
            peers: Vec::new(),
            runtime: HashMap::new(),
            rr_cursor: 0,
        }
    }

    pub fn set_peers<I>(&mut self, peers: I)
    where
        I: IntoIterator<Item = GatewayPeerDescriptor>,
    {
        self.peers = peers.into_iter().collect();
        let active_ids: HashSet<&str> = self.peers.iter().map(|peer| peer.agent_id.as_str()).collect();
        self.runtime.retain(|agent_id, _| active_ids.contains(agent_id.as_str()));

        let now_ms = self.now_ms();
        for peer in &self.peers {
            self.runtime
                .entry(peer.agent_id.clone())
                .or_insert_with(|| PeerRuntimeState::new(now_ms));
        }
    }

    pub fn update_peer_runtime_state(
        &mut self,
        agent_id: &str,
        state: Option<i32>,
        last_heartbeat_ms: Option<u64>,
        cooldown_until_ms: Option<u64>,
    ) {
        let now_ms = self.now_ms();
        let runtime = self
            .runtime
            .entry(agent_id.to_string())
            .or_insert_with(|| PeerRuntimeState::new(now_ms));

        if let Some(state) = state {
            runtime.state = state;
        }
        if let Some(last_heartbeat_ms) = last_heartbeat_ms {
            runtime.last_heartbeat_ms = last_heartbeat_ms;
        }
        if let Some(cooldown_until_ms) = cooldown_until_ms {
            runtime.cooldown_until_ms = cooldown_until_ms;
        }
    }

    pub fn touch_peer_heartbeat(&mut self, agent_id: &str, state: Option<i32>, now_ms: Option<u64>) {
        self.update_peer_runtime_state(
            agent_id,
            state,
            Some(now_ms.unwrap_or_else(|| self.now_ms())),
            None,
        );
    }

    pub fn record_peer_overloaded(
        &mut self,
        agent_id: &str,
        retry_after_ms: Option<u64>,
        local_cooldown_ms: Option<u64>,
    ) {
        let now_ms = self.now_ms();
        let runtime = self
            .runtime
            .entry(agent_id.to_string())
            .or_insert_with(|| PeerRuntimeState::new(now_ms));
        let cooldown_ms = retry_after_ms
            .unwrap_or(0)
            .max(local_cooldown_ms.unwrap_or(0));
        runtime.cooldown_until_ms = runtime.cooldown_until_ms.max(now_ms.saturating_add(cooldown_ms));
    }

    pub fn is_eligible(&self, peer: &GatewayPeerDescriptor) -> bool {
        if peer.registration_type != REGISTRATION_TYPE_SWARM_GATEWAY {
            return false;
        }
        if peer.agent_id == self.local_agent_id {
            return false;
        }
        same_capabilities(&peer.capabilities, &self.local_capabilities)
    }

    pub fn select_peer(&mut self) -> Option<String> {
        let healthy: Vec<&GatewayPeerDescriptor> =
            self.peers.iter().filter(|peer| self.is_healthy(peer)).collect();
        if healthy.is_empty() {
            return None;
        }

        let idx = self.rr_cursor % healthy.len();
        self.rr_cursor = self.rr_cursor.saturating_add(1);
        Some(healthy[idx].agent_id.clone())
    }

    fn is_healthy(&self, peer: &GatewayPeerDescriptor) -> bool {
        if !self.is_eligible(peer) {
            return false;
        }
        if !(self.peer_health_fn)(peer) {
            return false;
        }

        let Some(runtime) = self.runtime.get(&peer.agent_id) else {
            return false;
        };
        let now_ms = self.now_ms();
        if runtime.cooldown_until_ms > now_ms {
            return false;
        }
        if now_ms.saturating_sub(runtime.last_heartbeat_ms) > self.liveness_threshold_ms {
            return false;
        }
        if matches!(
            runtime.state,
            agent_state::INITIALIZING | agent_state::FAILED | agent_state::SHUTTING_DOWN
        ) {
            return false;
        }
        true
    }

    fn now_ms(&self) -> u64 {
        (self.now_ms_fn)()
    }
}

#[derive(Clone)]
pub struct GatewayRedirectEmitterOptions {
    pub agent_id: String,
    pub capabilities: Vec<String>,
    pub retry_after_ms: u64,
    pub peer_descriptors: Vec<GatewayPeerDescriptor>,
    pub now_ms_fn: Option<NowMsFn>,
    pub peer_health_fn: Option<PeerHealthFn>,
    pub peer_liveness_threshold_ms: u64,
}

impl GatewayRedirectEmitterOptions {
    pub fn new(agent_id: String, capabilities: Vec<String>) -> Self {
        Self {
            agent_id,
            capabilities,
            retry_after_ms: 1000,
            peer_descriptors: Vec::new(),
            now_ms_fn: None,
            peer_health_fn: None,
            peer_liveness_threshold_ms: DEFAULT_PEER_LIVENESS_THRESHOLD_MS,
        }
    }
}

pub struct GatewayRedirectEmitter {
    retry_after_ms: u64,
    peer_selector: PeerSelector,
}

impl GatewayRedirectEmitter {
    pub fn new(options: GatewayRedirectEmitterOptions) -> Self {
        let mut peer_selector = PeerSelector::new(PeerSelectorOptions {
            local_agent_id: options.agent_id,
            local_capabilities: options.capabilities,
            now_ms_fn: options.now_ms_fn,
            peer_health_fn: options.peer_health_fn,
            liveness_threshold_ms: options.peer_liveness_threshold_ms,
        });
        peer_selector.set_peers(options.peer_descriptors);

        Self {
            retry_after_ms: options.retry_after_ms,
            peer_selector,
        }
    }

    pub fn set_peer_descriptors<I>(&mut self, peers: I)
    where
        I: IntoIterator<Item = GatewayPeerDescriptor>,
    {
        self.peer_selector.set_peers(peers);
    }

    pub fn update_peer_runtime_state(
        &mut self,
        agent_id: &str,
        state: Option<i32>,
        last_heartbeat_ms: Option<u64>,
        cooldown_until_ms: Option<u64>,
    ) {
        self.peer_selector
            .update_peer_runtime_state(agent_id, state, last_heartbeat_ms, cooldown_until_ms);
    }

    pub fn touch_peer_heartbeat(&mut self, agent_id: &str, state: Option<i32>, now_ms: Option<u64>) {
        self.peer_selector.touch_peer_heartbeat(agent_id, state, now_ms);
    }

    pub fn record_peer_overloaded(
        &mut self,
        agent_id: &str,
        retry_after_ms: Option<u64>,
        local_cooldown_ms: Option<u64>,
    ) {
        self.peer_selector
            .record_peer_overloaded(agent_id, retry_after_ms, local_cooldown_ms);
    }

    pub fn emit_overloaded_response(&mut self, request: &HandoffRequest) -> HandoffResponse {
        let allow_spillover = request
            .delegation_policy
            .as_ref()
            .is_some_and(|policy| policy.allow_spillover_routing);

        if allow_spillover {
            if let Some(target) = self.peer_selector.select_peer() {
                return HandoffResponse {
                    accepted: false,
                    handoff_id: request.request_id.clone(),
                    rejection_reason: Some(
                        "Gateway at capacity; redirect to peer gateway".to_string(),
                    ),
                    accepting_agent: None,
                    rejection_code: Some(REJECTION_CODE_REDIRECT),
                    retry_after_ms: Some(0),
                    redirect_to_agent_id: Some(target),
                    status: HandoffStatus::Rejected,
                    metadata: HashMap::new(),
                };
            }
        }

        HandoffResponse {
            accepted: false,
            handoff_id: request.request_id.clone(),
            rejection_reason: Some("Gateway at capacity".to_string()),
            accepting_agent: None,
            rejection_code: Some(REJECTION_CODE_OVERLOADED),
            retry_after_ms: Some(self.retry_after_ms),
            redirect_to_agent_id: None,
            status: HandoffStatus::Rejected,
            metadata: HashMap::new(),
        }
    }
}

fn same_capabilities(peer_capabilities: &[String], local_capabilities: &HashSet<String>) -> bool {
    let peer_set: HashSet<&str> = peer_capabilities.iter().map(String::as_str).collect();
    if peer_set.len() != local_capabilities.len() {
        return false;
    }
    local_capabilities
        .iter()
        .all(|capability| peer_set.contains(capability.as_str()))
}

fn default_now_ms() -> u64 {
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_else(|_| Duration::from_millis(0));
    now.as_millis().min(u128::from(u64::MAX)) as u64
}
